packages/fxa-auth-server/lib/routes/utils/subscriptions.js

```diff
@@ -29,43 +29,131 @@ module.exports = {
     auth,
     db,
     uid,
-    client_id
+    client_id,
+    stripeHelper,
+    email
   ) {
-    const {
-      subscriptions: { productCapabilities = {}, clientCapabilities = {} } = {},
-    } = config;
-
-    const subscriptions = (await db.fetchAccountSubscriptions(uid)) || [];
-
-    const subscribedProducts = [
-      // All accounts get this product
-      PRODUCT_REGISTERED,
-      // Other products come from actual subscriptions
-      ...subscriptions.map(({ productId }) => productId),
-    ];
+    const capabilitiesToReveal = new Set();
+
+    // All accounts get this psuedo-product
+    const subscribedProducts = [PRODUCT_REGISTERED];
+
+    if (stripeHelper) {
+      await fetchSubscribedProductsFromStripe(
+        subscribedProducts,
+        uid,
+        stripeHelper,
+        email
+      );
+      await checkCapabilitiesFromStripe(
+        subscribedProducts,
+        capabilitiesToReveal,
+        client_id,
+        stripeHelper
+      );
+    } else {
+      // TODO: issue #3846 - remove this conditional branch
+      await fetchSubscribedProductsFromLocalDB(subscribedProducts, db, uid);
+    }
+
     // Accounts with at least one subscription get this product
-    if (subscriptions.length > 0) {
+    if (subscribedProducts.length > 1) {
       subscribedProducts.push(PRODUCT_SUBSCRIBED);
     }
 
-    const subscribedCapabilities = subscribedProducts.reduce(
-      (capabilities, product) =>
-        capabilities.concat(productCapabilities[product] || []),
-      []
-    );
-
-    const clientVisibleCapabilities = clientCapabilities[client_id] || [];
-
-    const capabilitiesToReveal = new Set(
-      subscribedCapabilities.filter(
-        capability =>
-          auth.strategy === 'sessionToken' ||
-          clientVisibleCapabilities.includes(capability)
-      )
+    // TODO: issue #3913 - remove support for capabilities from server config
+    checkCapabilitiesFromServerConfig(
+      capabilitiesToReveal,
+      subscribedProducts,
+      client_id,
+      config,
+      auth
     );
 
     return capabilitiesToReveal.size > 0
       ? Array.from(capabilitiesToReveal)
       : undefined;
   },
-};
+});
+
+async function fetchSubscribedProductsFromStripe(
```

Comment:
I don't understand why these functions all manipulate passed in arguments instead of constructing new return values.

Reply:
They all work in part on constructing an array, so I figured it was easier to change in-place rather than recreate at each stage